File: wk01/src/naive.py
def naive(string: str) -> list[int]:
    n = len(string)  # 6
    z = [0] * n
    z[0] = n

    for k in range(1, n):
        i = 0
        while k + i < n and string[i] == string[k + i]:
            i += 1
        z[k] = i

    return z


if __name__ == "__main__":
    z = naive("abxbab")
    print(z)
    assert z == [6, 0, 0, 0, 2, 0]

File: wk01/src/naive.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <printarr.h>
#include <assert.h>

int* naive(const char* string, size_t* out_n) {
    size_t n = strlen(string);
    int* z = calloc(n, sizeof(int));  // allocate on heap with zeros
    z[0] = n;

    for (int k = 1; k < n; k++) {
        size_t i = 0;
        while (k + i < n && string[i] == string[k + i]) {
            i++;
        }
        z[k] = i;
    }

    *out_n = n;  // writing length back to caller
    return z;
}

int main() {
    const char string[] = "abxbab";

    size_t n;
    int* z = naive(string, &n);

    printarr(z, n);
    int expected[] = {6,0,0,0,2,0};
    assert(memcmp(z, expected, n * sizeof(int)) == 0);

    free(z); // free stored z array from heap

    return 0;
}
